File: app/src/main/java/net/accelf/yuito/TimelineStreamingListener.java
package net.accelf.yuito;

import android.text.Spanned;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.keylesspalace.tusky.appstore.EventHub;
import com.keylesspalace.tusky.appstore.StatusDeletedEvent;
import com.keylesspalace.tusky.appstore.StreamUpdateEvent;
import com.keylesspalace.tusky.entity.Status;
import com.keylesspalace.tusky.entity.StreamEvent;
import com.keylesspalace.tusky.json.SpannedTypeAdapter;

import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

public class TimelineStreamingListener extends WebSocketListener {

    private Gson gson = buildGson();
    private boolean isFirstStatus = true;

    private EventHub eventHub;

    private static Gson buildGson() {
        return new GsonBuilder()
                .registerTypeAdapter(Spanned.class, new SpannedTypeAdapter())
                .create();
    }

    public TimelineStreamingListener(EventHub eventHub) {
        this.eventHub = eventHub;
    }

    @Override
    public void onOpen(WebSocket webSocket, Response response) {
        Log.d("StreamingListener", "Stream connected.");
    }

    @Override
    public void onMessage(WebSocket webSocket, String text) {
        StreamEvent event = gson.fromJson(text, StreamEvent.class);

        String payload = event.getPayload();
        switch (event.getEvent()) {
            case UPDATE:
                Status status = gson.fromJson(payload, Status.class);
                eventHub.dispatch(new StreamUpdateEvent(status, isFirstStatus));
                isFirstStatus = false;
                break;
            case DELETE:
                eventHub.dispatch(new StatusDeletedEvent(payload));
                break;
        }
    }

    @Override
    public void onClosed(WebSocket webSocket, int code, String reason) {
        Log.d("StreamingListener", "Stream closed.");
    }
}

File: app/src/main/java/net/accelf/yuito/TimelineStreamingListener.kt
package net.accelf.yuito

import android.util.Log
import com.google.gson.Gson
import com.keylesspalace.tusky.appstore.EventHub
import com.keylesspalace.tusky.appstore.PreferenceChangedEvent
import com.keylesspalace.tusky.appstore.StatusDeletedEvent
import com.keylesspalace.tusky.appstore.StreamUpdateEvent
import com.keylesspalace.tusky.components.timeline.TimelineViewModel
import com.keylesspalace.tusky.entity.Filter
import com.keylesspalace.tusky.entity.Status
import com.keylesspalace.tusky.entity.StreamEvent
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

class TimelineStreamingListener(
    private val eventHub: EventHub,
    private val gson: Gson,
    private val kind: TimelineViewModel.Kind,
    private val identifier: String? = null,
) : WebSocketListener() {

    private val target = if (identifier == null) { kind.name } else { kind.name + ":" + identifier }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        Log.d(TAG, "Stream connected to: $target")
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        val event = gson.fromJson(text, StreamEvent::class.java)
        val payload = event.payload
        when (event.event) {
            StreamEvent.EventType.UPDATE -> {
                val status = gson.fromJson(payload, Status::class.java)
                eventHub.dispatch(StreamUpdateEvent(status, kind, identifier))
            }
            StreamEvent.EventType.DELETE -> eventHub.dispatch(StatusDeletedEvent(payload))
            StreamEvent.EventType.FILTERS_CHANGED -> eventHub.dispatch(PreferenceChangedEvent(Filter.HOME)) // It may be not a home but it doesn't matter
            else -> Log.d(TAG, "Unsupported event type.")
        }
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        Log.d(TAG, "Stream closed for: $target")
    }

    companion object {
        private const val TAG = "StreamingListener"
    }
}
